### ISSUE TITLE
NewRoundTripper for injected http clients

### ISSUE THREAD

**Issue by ianlopshire** [newrelic/go-agent#79]
The current implementation of  `NewRoundTripper` makes it difficult to use when a `http.client` is passed into a service via dependency injection. Instead of the implementation being transaction specific, it should check the `http.request` context for a `rewrelic.Transaction`.

Consider this example of a very basic API client:

```go
type APIClient struct {
	instance string
	client   *http.Client
}

func NewAPIClient(instance string, httpClient *http.Client) *APIClient {
	return &APIClient{
		instance:   instance,
		httpClient: httpClient,
	}
}

func (c *APIClient) DoSomeAction(ctx context.Context) (res *http.Response, err error) {
	r := http.NewRequest("GET", c.instance, nil)
	r = r.WithContext(ctx)
	return c.client.Do(r)
}
```

It would be very helpful if the injected `http.Client` could be configured to automatically instrument requests. The desired usage would look something like:

```go
client := new(http.Client)
client.Transport = rewrelic.NewRoundTripper(c.Transport)

apiClient := NewAPIClient("http://example.com", client)
```

**Proposed NewRoundTripper Implementation:**

```go
func NewRoundTripper(original http.RoundTripper) http.RoundTripper {
	return roundTripperFunc(func(request *http.Request) (*http.Response, error) {
		txn := FromContext(request.Context())
		segment := StartExternalSegment(txn, request)

		if nil == original {
			original = http.DefaultTransport
		}
		response, err := original.RoundTrip(request)

		segment.Response = response
		segment.End()

		return response, err
	})
}
```

I'm fairly new to using the go agent so if there is an existing way to achieve what I've outlined, I'd be happy to hear about it. If not I'd be happy to submit a pull request after some discussion.

**Comment by willnewrelic** [comment]
Hi @ianlopshire 

We agree!  Great idea!  We plan to add this functionality in the next release.

For backwards compatibility, we plan on keeping the `NewRoundTripper` prototype the same:  Consumers will simply be encouraged to use `nil` for the `Transaction` parameter:

```go
client := new(http.Client)
client.Transport = rewrelic.NewRoundTripper(nil, c.Transport)
```

We also plan to update `StartExternalSegment` with similar behavior:  If transaction is nil, we will look in the request's context for a transaction.

**Comment by cgilling** [comment]
Great to see this, as the previous implementation seemed to make it pretty much impossible to cache connections

**Comment by willnewrelic** [comment]
This functionality has now been released in 2.2.0!

https://godoc.org/github.com/newrelic/go-agent#NewRoundTripper now takes a `nil` transaction.
